src/community/json-ld/src/main/java/org/geoserver/jsonld/builders/impl/DynamicValueBuilder.java

```diff
@@ -96,57 +112,48 @@ private Object evaluateExpressions(JsonBuilderContext context) {
 
-    private void prepareXpathFilter(Expression expr) {
-        List<Expression> params = ((Function) expr).getParameters();
-        if (params.size() > 0) {
-            for (Expression e : params) {
-                if (e instanceof XPathFunction) ((XPathFunction) e).setNamespaces(namespaces);
-                else if (e instanceof Function) {
-                    prepareXpathFilter(e);
+    private void literalXpathToPropertyName(Expression expr, String literalXpath) {
+        List<Expression> params = null;
+        if (expr instanceof Function) {
+            params = ((Function) expr).getParameters();
+        } else if (expr instanceof BinaryExpression) {
+            params =
+                    Arrays.asList(
+                            ((BinaryExpression) expr).getExpression1(),
+                            ((BinaryExpression) expr).getExpression2());
+        }
+        List<Expression> newParams = new ArrayList<>(params.size());
+        if (params != null || params.size() > 0) {
+            for (int i = 0; i < params.size(); i++) {
+                Expression e = params.get(i);
+                if (e instanceof Literal) {
+                    Literal literal = (Literal) params.get(i);
+                    String unquoted = literalXpath.replaceAll("'", "");
+                    if (String.valueOf(literal.getValue()).equals(unquoted)) {
+                        PropertyName pn = new AttributeExpressionImpl(unquoted, namespaces);
+                        newParams.add(i, pn);
+                    } else {
+                        newParams.add(i, e);
+                    }
+                } else if (e instanceof Function) {
+                    newParams.add(e);
+                    literalXpathToPropertyName(e, literalXpath);
                 }
             }
         }
-    }
-
-    /**
-     * Determines how many times is needed to walk up {@link JsonBuilderContext} in order to execute
-     * xpath, and cleans it from ../ notation.
-     *
-     * @param xpath
-     * @return
-     */
-    private String determineContextPos(String xpath) {
-        while (xpath.contains("../")) {
-            contextPos++;
-            xpath = xpath.replaceFirst("\\.\\./", "");
-        }
-        return xpath;
-    }
-
-    /**
-     * Extract xpath from a cql expression if present
-     *
-     * @param expression
-     * @return
-     */
-    private String workXpathFunction(String expression) {
-        // extract xpath from cql expression if present
-        int xpathI = expression.indexOf("xpath(");
-        if (xpathI != -1) {
-            int xpathI2 = expression.indexOf(")", xpathI);
-            String xpath = expression.substring(xpathI, xpathI2 + 1);
-            determineContextPos(xpath);
-            expression = expression.replaceAll("\\.\\./", "");
+        if (expr instanceof FunctionExpression) {
+            ((FunctionExpression) expr).setParameters(newParams);
+        } else if (expr instanceof MathExpressionImpl) {
```

Comment:
Here you probably want to check if its a binary expression and setting it accordingly no?

Reply:
Checking for MathExpressionImpl cause setParameters is available only casting to it. Anyway right now MathExpressionImpl is extended by all the implementation of BinaryExpression

Reply:
Interesting I didn't know that, thx @taba90!